import of isort package should be avoided when possible

### Steps to reproduce
1. call pylint.lint.Run() on Python 3.7

### Current behavior
`checkers/imports.py` imports isort, which in turn accesses pip internal packages, which in turn emits a Python DeprecationWarning.  This happens even though my Run() invocation does not enable the imports checkers.

```
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pylint/checkers/imports.py:41: in <module>
    import isort
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/isort/__init__.py:26: in <module>
    from .isort import SortImports  # noqa: F401
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/isort/isort.py:41: in <module>
    from .finders import FindersManager
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/isort/finders.py:24: in <module>
    from pip._internal.download import PipSession
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/__init__.py:40: in <module>
    from pip._internal.cli.autocompletion import autocomplete
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/cli/autocompletion.py:8: in <module>
    from pip._internal.cli.main_parser import create_main_parser
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/cli/main_parser.py:12: in <module>
    from pip._internal.commands import (
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/commands/__init__.py:6: in <module>
    from pip._internal.commands.completion import CompletionCommand
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/commands/completion.py:6: in <module>
    from pip._internal.cli.base_command import Command
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/cli/base_command.py:25: in <module>
    from pip._internal.index import PackageFinder
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_internal/index.py:14: in <module>
    from pip._vendor import html5lib, requests, six
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_vendor/html5lib/__init__.py:25: in <module>
    from .html5parser import HTMLParser, parse, parseFragment
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_vendor/html5lib/html5parser.py:8: in <module>
    from . import _tokenizer
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_vendor/html5lib/_tokenizer.py:16: in <module>
    from ._trie import Trie
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_vendor/html5lib/_trie/__init__.py:3: in <module>
    from .py import Trie as PyTrie
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_vendor/html5lib/_trie/py.py:6: in <module>
    from ._base import Trie as ABCTrie
../../../../virtualenv/python3.7.1/lib/python3.7/site-packages/pip/_vendor/html5lib/_trie/_base.py:3: in <module>
    from collections import Mapping

E           DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
```
### Expected behavior
Since `isort` is accessing unsupported pip internals, pylint should avoid importing isort unless the checker in question is run.  Since imports.py uses isort in only one place, the change seems straightforward.

In any case I've filed https://github.com/timothycrosley/isort/issues/924 on the isort side as well.

### pylint --version output
```
pylint 2.3.1
astroid 2.3.0
Python 3.7.0b3 (default, Aug 25 2018, 16:41:39)
```

--- a/doc/whatsnew/fragments/2866.performance
+++ b/doc/whatsnew/fragments/2866.performance
@@ -1,0 +1,7 @@
+Lazy-import ``isort`` so it is only loaded when import-ordering
+messages are enabled (``wrong-import-order``, ``ungrouped-imports``,
+``wrong-import-position``).  ``import isort`` alone takes ~60 ms,
+and lazy loading isort saves ~10% of ``pylint`` startup time when
+loading checkers from 220 ms to 190 ms.
+
+Closes #2866

--- a/doc/whatsnew/fragments/2866.performance
+++ b/doc/whatsnew/fragments/2866.performance
@@ -1,6 +1,6 @@
-Lazily import ``dill`` in ``pylint.lint.parallel`` and ``isort`` in
-``pylint.checkers.imports`` so they are only loaded when actually needed.
-``dill`` (~13 ms) is now deferred until parallel checking runs, and
-``isort`` (~60 ms) until import-ordering messages are checked.
+Lazily import ``isort``, ``dill``, ``multiprocessing``/``concurrent.futures``,
+and ``tomlkit`` so they are only loaded when actually needed.
+This reduces startup time by ~25% (e.g. ``--version``: 91 => 67 ms,
+``--help``: 176 => 133 ms, single-file lint: 272 => 226 ms).
 
 Closes #2866

--- a/doc/whatsnew/fragments/2866.performance
+++ b/doc/whatsnew/fragments/2866.performance
@@ -1,7 +1,6 @@
-Lazy-import ``isort`` so it is only loaded when import-ordering
-messages are enabled (``wrong-import-order``, ``ungrouped-imports``,
-``wrong-import-position``).  ``import isort`` alone takes ~60 ms,
-and lazy loading isort saves ~10% of ``pylint`` startup time when
-loading checkers from 220 ms to 190 ms.
+Lazily import ``dill`` in ``pylint.lint.parallel`` and ``isort`` in
+``pylint.checkers.imports`` so they are only loaded when actually needed.
+``dill`` (~13 ms) is now deferred until parallel checking runs, and
+``isort`` (~60 ms) until import-ordering messages are checked.
 
 Closes #2866

--- a/pylint/checkers/imports.py
+++ b/pylint/checkers/imports.py
@@ -15,9 +15,11 @@ from collections.abc import ItemsView, Sequence
 from functools import cached_property
 from typing import TYPE_CHECKING, Any
 
+if TYPE_CHECKING:
+    import isort
+
 import astroid
 import astroid.modutils
-import isort
 from astroid import nodes
 from astroid.nodes._base_nodes import ImportNode
 
@@ -778,6 +780,8 @@ class ImportsChecker(DeprecatedMixin, BaseChecker):
 
         Only valid after CLI parsing finished, i.e. not in __init__
         """
+        import isort  # pylint: disable=import-outside-toplevel
+
         return isort.Config(
             # There is no typo here. EXTRA_standard_library is
             # what most users want. The option has been named
@@ -788,7 +792,9 @@ class ImportsChecker(DeprecatedMixin, BaseChecker):
             known_third_party=self.linter.config.known_third_party,
         )
 
-    def _check_imports_order(self, _module_node: nodes.Module) -> tuple[
+    def _check_imports_order(  # pylint: disable=too-many-statements
+        self, _module_node: nodes.Module
+    ) -> tuple[
         list[tuple[ImportNode, str]],
         list[tuple[ImportNode, str]],
         list[tuple[ImportNode, str]],
@@ -797,6 +803,8 @@ class ImportsChecker(DeprecatedMixin, BaseChecker):
 
         Imports must follow this order: standard, 3rd party, 1st party, local
         """
+        import isort  # pylint: disable=import-outside-toplevel
+
         std_imports: list[tuple[ImportNode, str]] = []
         third_party_imports: list[tuple[ImportNode, str]] = []
         first_party_imports: list[tuple[ImportNode, str]] = []

--- a/pylint/config/arguments_manager.py
+++ b/pylint/config/arguments_manager.py
@@ -14,8 +14,6 @@ import warnings
 from collections.abc import Sequence
 from typing import TYPE_CHECKING, Any, TextIO
 
-import tomlkit
-
 from pylint import utils
 from pylint.config.argument import (
     _Argument,
@@ -302,6 +300,8 @@ class _ArgumentsManager:
         """Write a configuration file according to the current configuration into
         stdout.
         """
+        import tomlkit  # pylint: disable=import-outside-toplevel
+
         toml_doc = tomlkit.document()
         tool_table = tomlkit.table(is_super_table=True)
         toml_doc.add(tomlkit.key("tool"), tool_table)

--- a/pylint/lint/parallel.py
+++ b/pylint/lint/parallel.py
@@ -9,8 +9,6 @@ from collections import defaultdict
 from collections.abc import Iterable, Sequence
 from typing import TYPE_CHECKING, Any
 
-import dill
-
 from pylint import reporters
 from pylint.lint.utils import _augment_sys_path
 from pylint.message import Message
@@ -44,6 +42,8 @@ def _worker_initialize(
     :param extra_packages_paths: Extra entries to be added to `sys.path`
     """
     global _worker_linter  # pylint: disable=global-statement
+    import dill  # pylint: disable=import-outside-toplevel
+
     _worker_linter = dill.loads(linter)
     assert _worker_linter
 
@@ -144,6 +144,8 @@ def check_parallel(
     initializer = functools.partial(
         _worker_initialize, extra_packages_paths=extra_packages_paths
     )
+    import dill  # pylint: disable=import-outside-toplevel
+
     with ProcessPoolExecutor(
         max_workers=jobs, initializer=initializer, initargs=(dill.dumps(linter),)
     ) as executor:

--- a/pylint/lint/parallel.py
+++ b/pylint/lint/parallel.py
@@ -15,17 +15,8 @@ from pylint.message import Message
 from pylint.typing import FileItem
 from pylint.utils import LinterStats, merge_stats
 
-try:
-    import multiprocessing
-except ImportError:
-    multiprocessing = None  # type: ignore[assignment]
-
-try:
-    from concurrent.futures import ProcessPoolExecutor
-except ImportError:
-    ProcessPoolExecutor = None  # type: ignore[assignment,misc]
-
 if TYPE_CHECKING:
+
     from pylint.lint import PyLinter
 
 # PyLinter object used by worker processes when checking files using parallel mode
@@ -79,6 +70,8 @@ def _worker_check_single_file(
     int,
     defaultdict[str, list[Any]],
 ]:
+    import multiprocessing  # pylint: disable=import-outside-toplevel
+
     if not _worker_linter:
         raise RuntimeError("Worker linter not yet initialised")
     _worker_linter.open()
@@ -144,6 +137,9 @@ def check_parallel(
     initializer = functools.partial(
         _worker_initialize, extra_packages_paths=extra_packages_paths
     )
+    # pylint: disable-next=import-outside-toplevel
+    from concurrent.futures import ProcessPoolExecutor
+
     import dill  # pylint: disable=import-outside-toplevel
 
     with ProcessPoolExecutor(

--- a/pylint/lint/run.py
+++ b/pylint/lint/run.py
@@ -25,17 +25,6 @@ from pylint.lint.base_options import _make_run_options
 from pylint.lint.pylinter import MANAGER, PyLinter
 from pylint.reporters.base_reporter import BaseReporter
 
-try:
-    import multiprocessing
-    from multiprocessing import synchronize  # noqa pylint: disable=unused-import
-except ImportError:
-    multiprocessing = None  # type: ignore[assignment]
-
-try:
-    from concurrent.futures import ProcessPoolExecutor
-except ImportError:
-    ProcessPoolExecutor = None  # type: ignore[assignment,misc]
-
 
 def _query_cpu() -> int | None:
     """Try to determine number of CPUs allotted in a docker container.
@@ -110,10 +99,13 @@ def _cpu_count() -> int:
     # pylint: disable=not-callable,using-constant-test,useless-suppression
     if sched_getaffinity:
         cpu_count = len(sched_getaffinity(0))
-    elif multiprocessing:
-        cpu_count = multiprocessing.cpu_count()
     else:
-        cpu_count = 1
+        try:
+            import multiprocessing  # pylint: disable=import-outside-toplevel
+
+            cpu_count = multiprocessing.cpu_count()
+        except ImportError:
+            cpu_count = 1
     if sys.platform == "win32":
         # See also https://github.com/python/cpython/issues/94242
         cpu_count = min(cpu_count, 56)  # pragma: no cover
@@ -217,13 +209,16 @@ group are mutually exclusive.",
             )
             sys.exit(32)
         if linter.config.jobs > 1 or linter.config.jobs == 0:
-            if ProcessPoolExecutor is None:
+            try:
+                # pylint: disable-next=import-outside-toplevel,unused-import
+                from concurrent.futures import ProcessPoolExecutor  # noqa: F401
+            except ImportError:
                 print(
                     "concurrent.futures module is missing, fallback to single process",
                     file=sys.stderr,
                 )
                 linter.set_option("jobs", 1)
-            elif linter.config.jobs == 0:
+            if linter.config.jobs == 0:
                 linter.config.jobs = _cpu_count()
 
         if self._output:

